File: logitrack-mobile/lib/features/home/data/models/trip_record.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// โมเดลตาม shared-docs/schemas/trip-record.ts (TripRecords collection)
/// ใช้สำหรับบันทึกรับงาน (Loading Phase) และสถานะ trip อื่นๆ
class TripRecord {
  final String? id;
  final String status;
  final String jobType;
  final List<TripPhoto> photos;
  final TripOcrData? ocrData;
  final String? spxTripId;
  final String? taskId;
  final String? origin;
  final String? destination;
  final String? sealCode;
  final String? partnerCode;
  final String? driverId;
  final String? distance;
  final int? parcelCount;
  final String? sealTime;
  final String? totalWeight;
  final double? lat;
  final double? lng;
  final DateTime? deliveredTimestamp;
  final double? deliveredLat;
  final double? deliveredLng;

  /// STD = Scheduled Time of Departure (CreateAt / บันทึกเมื่อ)
  final DateTime? std;

  /// STA = Scheduled Time of Arrival (เวลา Seal รถ + durationMinutes จาก hub_soc_distances)
  final DateTime? sta;

  /// ATA = Actual Time of Arrival (จะอัปเดตเมื่อถึง Hub/SOC ด้วย Geofencing)
  final DateTime? ata;

  /// เวลาเดินทาง (นาที) จาก hub_soc_distances
  final double? durationMinutes;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final bool needsAdminReview;
  final String? reviewStatus;
  final String? reviewReason;
  final DateTime? resubmittedAt;

  static DateTime? _parseDate(dynamic v) {
    if (v == null) return null;
    if (v is DateTime) return v;
    if (v is Timestamp) return v.toDate();
    if (v is String) return DateTime.tryParse(v);
    return null;
  }

  /// สร้างจาก Firestore doc (หรือ map จาก API)
  factory TripRecord.fromMap(Map<String, dynamic> map, {String? id}) {
    final photosRaw = map['photos'];
    List<TripPhoto> photos = const [];
    if (photosRaw is List) {
      photos = photosRaw.map((e) {
        if (e is Map<String, dynamic>) {
          TripPhotoGeocoding? geo;
          final g = e['geocoding'];
          if (g is Map<String, dynamic>) {
            geo = TripPhotoGeocoding(
              lat: (g['lat'] as num?)?.toDouble(),
              lng: (g['lng'] as num?)?.toDouble(),
              address: g['address'] as String?,
              timestamp: _parseDate(g['timestamp']),
            );
          }
          return TripPhoto(
            url: (e['url'] as String?) ?? '',
            type: (e['type'] as String?) ?? '',
            geocoding: geo,
          );
        }
        return const TripPhoto(url: '', type: '');
      }).toList();
    }
    TripOcrData? ocrData;
    final ocr = map['ocrData'];
    if (ocr is Map<String, dynamic>) {
      ocrData = TripOcrData(
        tripId: ocr['tripId'] as String?,
        sealCode: ocr['sealCode'] as String?,
        secondarySealCode: ocr['secondarySealCode'] as String?,
        routeInfo: ocr['routeInfo'] as String?,
        partnerCode: ocr['partnerCode'] as String?,
        supplierCode: ocr['supplierCode'] as String?,
        releaseTime: ocr['releaseTime'] as String?,
        sealSource: ocr['sealSource'] as String?,
        ocrProfile: ocr['ocrProfile'] as String?,
      );
    }
    return TripRecord(
      id: id ?? map['id'] as String?,
      status: (map['status'] as String?) ?? 'loading',
      jobType: (map['jobType'] as String?) ?? 'first_mile',
      photos: photos,
      ocrData: ocrData,
      spxTripId: map['spxTripId'] as String?,
      taskId: map['taskId'] as String?,
      origin: map['origin'] as String?,
      destination: map['destination'] as String?,
      sealCode: map['sealCode'] as String?,
      partnerCode: map['partnerCode'] as String?,
      driverId: map['driverId'] as String?,
      distance: map['distance'] as String?,
      parcelCount: (map['parcelCount'] as num?)?.toInt(),
      sealTime: map['sealTime'] as String?,
      totalWeight: map['totalWeight'] as String?,
      lat: (map['lat'] as num?)?.toDouble(),
      lng: (map['lng'] as num?)?.toDouble(),
      deliveredTimestamp: _parseDate(map['deliveredTimestamp']),
      deliveredLat: (map['deliveredLat'] as num?)?.toDouble(),
      deliveredLng: (map['deliveredLng'] as num?)?.toDouble(),
      std: _parseDate(map['std']),
      sta: _parseDate(map['sta']),
      ata: _parseDate(map['ata']),
      durationMinutes: (map['durationMinutes'] as num?)?.toDouble(),
      createdAt: _parseDate(map['createdAt']),
      updatedAt: _parseDate(map['updatedAt']),
      needsAdminReview: map['needsAdminReview'] == true,
      reviewStatus: map['reviewStatus'] as String?,
      reviewReason: map['reviewReason'] as String?,
      resubmittedAt: _parseDate(map['resubmittedAt']),
    );
  }

  const TripRecord({
    this.id,
    required this.status,
    required this.jobType,
    this.photos = const [],
    this.ocrData,
    this.spxTripId,
    this.taskId,
    this.origin,
    this.destination,
    this.sealCode,
    this.partnerCode,
    this.driverId,
    this.distance,
    this.parcelCount,
    this.sealTime,
    this.totalWeight,
    this.lat,
    this.lng,
    this.deliveredTimestamp,
    this.deliveredLat,
    this.deliveredLng,
    this.std,
    this.sta,
    this.ata,
    this.durationMinutes,
    this.createdAt,
    this.updatedAt,
    this.needsAdminReview = false,
    this.reviewStatus,
    this.reviewReason,
    this.resubmittedAt,
  });

  Map<String, dynamic> toFirestore() {
    final now = DateTime.now();
    return {
      'status': status,
      'jobType': jobType,
      'photos': photos.map((p) => p.toMap()).toList(),
      if (ocrData != null) 'ocrData': ocrData!.toMap(),
      if (spxTripId != null) 'spxTripId': spxTripId,
      if (taskId != null) 'taskId': taskId,
      if (origin != null) 'origin': origin,
      if (destination != null) 'destination': destination,
      if (sealCode != null) 'sealCode': sealCode,
      if (partnerCode != null) 'partnerCode': partnerCode,
      if (driverId != null) 'driverId': driverId,
      if (distance != null) 'distance': distance,
      if (parcelCount != null) 'parcelCount': parcelCount,
      if (sealTime != null) 'sealTime': sealTime,
      if (totalWeight != null) 'totalWeight': totalWeight,
      if (lat != null) 'lat': lat,
      if (lng != null) 'lng': lng,
      if (deliveredTimestamp != null) 'deliveredTimestamp': deliveredTimestamp,
      if (deliveredLat != null) 'deliveredLat': deliveredLat,
      if (deliveredLng != null) 'deliveredLng': deliveredLng,
      if (std != null) 'std': std,
      if (sta != null) 'sta': sta,
      if (ata != null) 'ata': ata,
      if (durationMinutes != null) 'durationMinutes': durationMinutes,
      'createdAt': createdAt ?? now,
      'updatedAt': updatedAt ?? now,
      'needsAdminReview': needsAdminReview,
      if (reviewStatus != null) 'reviewStatus': reviewStatus,
      if (reviewReason != null) 'reviewReason': reviewReason,
      if (resubmittedAt != null) 'resubmittedAt': resubmittedAt,
    };
  }
}

class TripPhoto {
  final String url;
  final String type;
  final TripPhotoGeocoding? geocoding;

  const TripPhoto({required this.url, required this.type, this.geocoding});

  Map<String, dynamic> toMap() => {
    'url': url,
    'type': type,
    if (geocoding != null) 'geocoding': geocoding!.toMap(),
  };
}

class TripPhotoGeocoding {
  final double? lat;
  final double? lng;
  final String? address;
  final DateTime? timestamp;

  const TripPhotoGeocoding({this.lat, this.lng, this.address, this.timestamp});

  Map<String, dynamic> toMap() => {
    if (lat != null) 'lat': lat,
    if (lng != null) 'lng': lng,
    if (address != null) 'address': address,
    if (timestamp != null) 'timestamp': timestamp?.toIso8601String(),
  };
}

class TripOcrData {
  final String? tripId;
  final String? sealCode;
  final String? secondarySealCode;
  final String? routeInfo;
  final String? partnerCode;
  final String? supplierCode;
  final String? releaseTime;
  final String? sealSource;
  final String? ocrProfile;

  const TripOcrData({
    this.tripId,
    this.sealCode,
    this.secondarySealCode,
    this.routeInfo,
    this.partnerCode,
    this.supplierCode,
    this.releaseTime,
    this.sealSource,
    this.ocrProfile,
  });

  Map<String, dynamic> toMap() => {
    if (tripId != null) 'tripId': tripId,
    if (sealCode != null) 'sealCode': sealCode,
    if (secondarySealCode != null) 'secondarySealCode': secondarySealCode,
    if (routeInfo != null) 'routeInfo': routeInfo,
    if (partnerCode != null) 'partnerCode': partnerCode,
    if (supplierCode != null) 'supplierCode': supplierCode,
    if (releaseTime != null) 'releaseTime': releaseTime,
    if (sealSource != null) 'sealSource': sealSource,
    if (ocrProfile != null) 'ocrProfile': ocrProfile,
  };
}

File: shared-docs/schemas/trip-record.ts
import { z } from "zod";

// --- Status (Driver workflow state machine) ---
export const TRIP_STATUS_ENUM = [
    "loading",
    "departure",
    "in_transit",
    "incident",
    "delivered",
] as const;

// --- Job Type (auto-detected from origin: Hub = first_mile, SOC = line_haul) ---
export const TRIP_JOB_TYPE_ENUM = ["FIRST_MILE", "LINE_HAUL"] as const;

// --- Photo type (Loading: pre-close, closing, seal, runsheet + extra handover pages; Delivery: pre_open, opening, empty_container, runsheet_received) ---
export const TRIP_PHOTO_TYPE_ENUM = [
    "pre_close",
    "closing",
    "seal",
    "runsheet",
    "runsheet_extra_1",
    "runsheet_extra_2",
    "runsheet_extra_3",
    "pre_open",
    "opening",
    "empty_container",
    "runsheet_received",
] as const;

// --- Geocoding metadata (overlay on in-app camera photos) ---
export const tripPhotoGeocodingSchema = z.object({
    lat: z.number().optional(),
    lng: z.number().optional(),
    address: z.string().optional(),
    timestamp: z.date().optional().or(z.string().optional()), // Overlay timestamp
});

// --- Single photo entry (URL, type, geocoding) ---
export const tripPhotoSchema = z.object({
    url: z.string().url("Photo URL must be valid"),
    type: z.enum(TRIP_PHOTO_TYPE_ENUM),
    geocoding: tripPhotoGeocodingSchema.optional(),
});

// --- OCR data from Shopee screenshot (tripId LTQ..., sealCode SPX..., route info) ---
export const tripOcrDataSchema = z.object({
    tripId: z.string().optional(), // Pattern: LTQ...
    sealCode: z.string().optional(), // Pattern: SPX...
    secondarySealCode: z.string().optional(),
    routeInfo: z.string().optional(), // Extracted from Shopee screenshot
    partnerCode: z.string().optional(), // e.g. TTP, JWT from LH-XXX / FM-XXX on runsheet OCR
    supplierCode: z.string().optional(),
    releaseTime: z.string().optional(),
    sealSource: z.enum(["scanned", "ocr_fallback"]).optional(),
    ocrProfile: z.enum(["spx_runsheet", "zx_waybill", "seal"]).optional(),
});

// --- Main Trip Record schema (TripRecords collection - SSOT for Web Dashboard & Billing) ---
export const tripRecordSchema = z.object({
    id: z.string().optional(),

    status: z.enum(TRIP_STATUS_ENUM).default("loading"),
    jobType: z.enum(TRIP_JOB_TYPE_ENUM),

    photos: z.array(tripPhotoSchema).optional().default([]),

    ocrData: tripOcrDataSchema.optional(),

    // Cross-reference (e.g. First Mile task, SPX Runsheet)
    spxTripId: z.string().optional(),
    taskId: z.string().optional(),

    // ต้นทาง / ปลายทาง (จากฟอร์มบันทึกรับงาน)
    origin: z.string().optional(),
    destination: z.string().optional(),

    // Seal code (จากฟอร์มหรือ OCR)
    sealCode: z.string().optional(),

    // ช่องทาง / พาร์ทเนอร์มอบงาน (จาก OCR รันชีท LH-XXX / FM-XXX)
    partnerCode: z.string().optional(),

    // Delivery completion (เมื่อส่งงานเสร็จ)
    deliveredTimestamp: z.date().optional(),
    deliveredLat: z.number().optional(),
    deliveredLng: z.number().optional(),

    // Billing snapshot (optional; captured when delivered to keep invoice numbers stable)
    billingEstimateThb: z.number().optional(),
    billingBaseRateThb: z.number().optional(),
    billingRateImportId: z.string().optional(),
    billingLookupHubId: z.string().optional(),
    billingLookupDestination: z.string().optional(),
    billingFuelAdjustmentId: z.string().optional(),
    billingRateMultiplier: z.number().optional(),
    billingAddThbPerTrip: z.number().optional(),
    /** Effective date of the *fuel adjustment* that applied — not of the rate round. */
    billingEffectiveFromDateStr: z.string().optional(),
    billingCustomerId: z.string().optional(),
    // Multi-delivery billing snapshot (only when isMultiDelivery = true)
    billingStopChargeThb: z.number().optional(),
    billingIsMultiDelivery: z.boolean().optional(),
    billingMultiDeliveryBreakdown: z
        .array(
            z.object({
                stopIndex: z.number(),
                destination: z.string(),
                baseRateThb: z.number(),
                finalRateThb: z.number(),
            })
        )
        .optional(),
    // Rate round + fuel band provenance (ADR 0009 §4) — denormalized at compute time so an
    // invoice never has to follow billingFuelAdjustmentId into a row that may have moved.
    /** `yyyy-MM-dd` the price of this trip last changed — the later of rate-entry / fuel-adjustment. */
    billingRoundEffectiveFromDateStr: z.string().optional(),
    /** Inclusive lower bound of the ฿1.00 diesel band, e.g. 41.01 for the band 41.01–42.00. */
    billingFuelBandLowerThb: z.number().optional(),
    /** Inclusive upper bound of the same band, e.g. 42. */
    billingFuelBandUpperThb: z.number().optional(),
    /** The announced retail diesel price the band was derived from. */
    billingReferenceFuelPriceThb: z.number().optional(),

    // Audit
    createdAt: z.date().optional(),
    updatedAt: z.date().optional(),
});

export type TripStatus = (typeof TRIP_STATUS_ENUM)[number];
export type TripJobType = (typeof TRIP_JOB_TYPE_ENUM)[number];
export type TripPhotoType = (typeof TRIP_PHOTO_TYPE_ENUM)[number];
export type TripPhotoGeocoding = z.infer<typeof tripPhotoGeocodingSchema>;
export type TripPhoto = z.infer<typeof tripPhotoSchema>;
export type TripOcrData = z.infer<typeof tripOcrDataSchema>;
export type TripRecord = z.infer<typeof tripRecordSchema>;
